cpp/include/raft/sparse/solver/detail/lanczos.cuh

```diff
@@ -1396,4 +1438,658 @@ int computeLargestEigenvectors(
+
+  auto cusparse_h = resource::get_cusparse_handle(handle);
+  cusparseSpMatDescr_t cusparse_A;
+  raft::sparse::detail::cusparsecreatecsr(&cusparse_A,
+                                          A->nrows_,
+                                          A->ncols_,
+                                          A->nnz_,
+                                          const_cast<index_type_t*>(A->row_offsets_),
+                                          const_cast<index_type_t*>(A->col_indices_),
+                                          const_cast<value_type_t*>(A->values_));
+
+  cusparseDnVecDescr_t cusparse_v;
+  cusparseDnVecDescr_t cusparse_u;
+  raft::sparse::detail::cusparsecreatednvec(&cusparse_v, n, v.data_handle());
+  raft::sparse::detail::cusparsecreatednvec(&cusparse_u, n, u.data_handle());
+
+  value_type_t one  = 1;
+  value_type_t zero = 0;
+  size_t bufferSize;
+  raft::sparse::detail::cusparsespmv_buffersize(cusparse_h,
+                                                CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                                &one,
+                                                cusparse_A,
+                                                cusparse_v,
+                                                &zero,
+                                                cusparse_u,
+                                                CUSPARSE_SPMV_ALG_DEFAULT,
+                                                &bufferSize,
+                                                stream);
+  auto cusparse_spmv_buffer = raft::make_device_vector<value_type_t>(handle, bufferSize);
+
+  // LOOP
+  for (int i = start_idx; i < end_idx; i++) {
+    raft::sparse::detail::cusparsespmv(cusparse_h,
+                                       CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                       &one,
+                                       cusparse_A,
+                                       cusparse_v,
+                                       &zero,
+                                       cusparse_u,
+                                       CUSPARSE_SPMV_ALG_DEFAULT,
+                                       cusparse_spmv_buffer.data_handle(),
+                                       stream);
+
+    auto alpha_i  = raft::make_device_scalar_view(&alpha(0, i));
+    auto v_vector = raft::make_device_vector_view<const value_type_t>(v.data_handle(), n);
+    auto u_vector = raft::make_device_vector_view<const value_type_t>(u.data_handle(), n);
+    raft::linalg::dot(handle, v_vector, u_vector, alpha_i);
+
+    raft::matrix::fill(handle, vv, zero);
+
+    auto cublas_h = resource::get_cublas_handle(handle);
+
+    value_type_t alpha_i_host = 0;
+    value_type_t b            = 0;
+    value_type_t mone         = -1;
+
+    raft::copy<value_type_t>(&b, &beta(0, (i - 1 + ncv) % ncv), 1, stream);
+    raft::copy<value_type_t>(&alpha_i_host, &(alpha(0, i)), 1, stream);
+
+    raft::linalg::detail::cublasaxpy(
```

Comment:
Please use the publicly available APIs and don't call into detail for these wrappers. The whole reason these wrappers exist in detail is that the private bits can be hidden and exposed through a single public interface. When you call into these directly, you are side-stepping that interface. You should be calling `raft::linalg::saxpy` (e.g. from your perspective, it should not matter that this is using cublas at all.. it should just be a raft functions. Please do the same with all of these calls into the `_wrapper` headers.